### DISCUSSION TITLE
Store return undefined when calling setter function inside an object

### DISCUSSION THREAD

**Discussion by khake19** [pmndrs/zustand#1904] [Q&A] [answered]
I have a setFilters defined inside an object:
```js
const userStore = create(persist({
  filters: {
    matter: {
      filters: [],
      setFilters: (filters: any) => {
        set((state) => {
          return { ...state, filters };
        });
      },
    },
    task: {
      filters: [],
      setFilters: (filters: any) => {
        set((state) => {
          return { ...state, filters };
        });
      },
    }
  },
  setSomething: () => set(state => ({something: 1 })
}), 
{ name: 'persistStore' }
)
```
When refreshing the browser and calling the store `setFilters` it returned undefined while `filters return correctly,
This is only happening when using persist middleware, and setter function outside object works perfectly fined.

```js
const setFilters = userStore((state) => state.filters.matter.setFilters); // return undefined;
const filters = userStore((state) => state.filters.matter.filters); // return [];
const setSomething = userStore((state) => state.setSomething); // return func;
```

Why persisting state also remove the setter method inside of that state?

**Comment by dbritto-dev** [comment, marked as answer]
@khake19 that's expected since you can not serialize/deserialize a function

**Reply by khake19** [reply]
Hi @dbritto-dev , Thanks for the response, I have now fixed this by combining `persisted state` and `current state`

```js
merge: (persistedState, currentState) => {
 return merge(persistedState, currentState);
},
```

**Reply by ronnizan** [reply]
@khake19 can you explain how do you solved it? I'm facing the same issue

**Reply by joshuarichards001** [reply]
For anyone stumbling across this in the future I used [deepmerge](https://www.npmjs.com/package/deepmerge) but you could use any deep merging function. Essentially my problem was that the default merge function was clearing all of the nested functions I had. All I had to do to fix the problem was add this code with `deepmerge` into the persist settings:

```javascript
{
  name: "persist-store-name",
  merge: (persistedState, currentState) =>
    deepmerge(currentState, persistedState as AppState),
}
```

Hope that helps!

**Reply by dbritto-dev** [reply]
btw, I encourage you to read the new api docs the new examples and docs are under middlewares, apis and hooks the old ones still a bit outdated
